metals/src/main/scala/scala/meta/internal/metals/FallbackDefinitionProvider.scala

```diff
@@ -0,0 +1,155 @@
+package scala.meta.internal.metals
+
+import scala.meta.Term
+import scala.meta.Type
+import scala.meta._
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.mtags
+import scala.meta.internal.mtags.GlobalSymbolIndex
+import scala.meta.internal.parsing.Trees
+import scala.meta.io.AbsolutePath
+import scala.meta.tokens.Token
+
+import org.eclipse.lsp4j.Location
+import org.eclipse.lsp4j.Position
+import org.eclipse.lsp4j.{Range => LspRange}
+
+class FallbackDefinitionProvider(
+    trees: Trees,
+    index: GlobalSymbolIndex,
+) {
+
+  /**
+   *  Tries to find an identifier token at the current position to
+   *  guess the symbol to find and then searches for it in the symbol index.
+   *  This is the last possibility for finding the definition.
+   *
+   * @param path path of the current file
+   * @param pos position we are searching for
+   * @return possible definition locations based on exact symbol search
+   */
+  def search(
+      path: AbsolutePath,
+      pos: Position,
+      isScala3: Boolean,
+  ): Option[DefinitionResult] = {
+    val range = new LspRange(pos, pos)
+
+    val defResult = for {
+      tokens <- trees.tokenized(path)
+      ident <- tokens.collectFirst {
+        case id: Token.Ident if id.pos.encloses(range) => id
+      }
+      tree <- trees.get(path)
+    } yield {
+      lazy val nameTree = trees.findLastEnclosingAt(path, pos)
+
+      // for sure is not a class/trait/enum if we access it via select
+      lazy val isInSelectPosition =
+        nameTree.flatMap(_.parent).exists(isInSelect(_, range))
+
+      lazy val isInTypePosition = nameTree.exists(_.is[Type.Name])
+
+      def guessObjectOrClass(parts: List[String]) = {
+        val symbolPrefix = mtags.Symbol
+          .guessSymbolFromParts(parts, isScala3)
+          .value
+        if (isInSelectPosition) List(symbolPrefix + ".")
+        else if (isInTypePosition) List(symbolPrefix + "#")
+        else List(".", "#").map(ending => symbolPrefix + ending)
+      }
+
+      // Get all select parts to build symbol from it later
+      val proposedNameParts =
+        nameTree
+          .flatMap(_.parent)
+          .map {
+            case tree: Term.Select if nameTree.contains(tree.name) =>
+              nameFromSelect(tree, Nil)
+            case _ => List(ident.value)
+          }
+          .getOrElse(List(ident.value))
+
+      val proposedCurrentPackageSymbol =
+        guessObjectOrClass(
+          trees
+            .packageAtPosition(path, pos)
+            .getOrElse("_empty_") +: proposedNameParts
+        )
```

Comment:
You're not handling package objects and things like:
```Scala
object O {
  class A
  val i: A@@ = ???
}
```
Also, in:
```Scala
package a
package b
```
a.<..> should also be findable.

But here you should prefer the inner scope instead of finding all...

Reply:
The first thing should be found by the presentation compiler without an issue, if it's not finding it the code might be too broken.

I added handling of the second case with a test, though it looks like this is also currently mostly handled by the presentation compiler correctly, but I checked manually anyway to see if the results are the same. We could have a test suite that tries to compare both, but I think it is enough to know it's working for now.